File: Python/trimBST.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: trimBST.py
@Function: 修剪二叉搜索树
@Link: https://leetcode-cn.com/problems/trim-a-binary-search-tree/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-01
"""


class Solution(object):
    def trimBST(self, root, L, R):
        def trim(node):
            if not node:
                return None
            elif node.val > R:
                return trim(node.left)
            elif node.val < L:
                return trim(node.right)
            else:
                node.left = trim(node.left)
                node.right = trim(node.right)
                return node

        return trim(root)


# ----------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/trimBST.cpp
/**
 * @File    : trimBST.cpp
 * @Brief   : 修剪二叉搜索树
 * @Link    : https://leetcode-cn.com/problems/trim-a-binary-search-tree/
 * @Author  : Wei Li
 * @Date    : 2021-08-01
*/

#include <iostream>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *trimBST(TreeNode *root, int L, int R)
    {
        if (!root)
        {
            return root;
        }
        if (root->val > R)
        {
            return trimBST(root->left, L, R);
        }
        if (root->val < L)
        {
            return trimBST(root->right, L, R);
        }
        root->left = trimBST(root->left, L, R);
        root->right = trimBST(root->right, L, R);
        return root;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
